source/Octopus.Tentacle.Client/Retries/RpcCallRetryHandler.cs

```diff
@@ -41,21 +41,31 @@ public RpcCallRetryHandler(TimeSpan retryTimeout, TimeoutStrategy timeoutStrateg
         {
             Exception? lastException = null;
             var started = new Stopwatch();
+            var nextSleepDuration = TimeSpan.Zero;
+            var remainingRetryDurationBuffer = TimeSpan.FromSeconds(1);
+            var hasRetried = false;
 
             async Task OnRetryAction(Exception exception, TimeSpan sleepDuration, int retryCount, Context context)
             {
                 lastException = exception;
+                nextSleepDuration = sleepDuration;
                 var elapsedDuration = started.Elapsed;
+                var remainingRetryDuration = RetryTimeout - elapsedDuration - sleepDuration;
 
-                if (onRetryAction != null)
+                if (ShouldRetry(remainingRetryDuration))
                 {
-                    await onRetryAction.Invoke(exception, sleepDuration, retryCount, RetryTimeout, elapsedDuration, cancellationToken).ConfigureAwait(false);
+                    hasRetried = true;
+
+                    if (onRetryAction != null)
+                    {
+                        await onRetryAction.Invoke(exception, sleepDuration, retryCount, RetryTimeout, elapsedDuration, cancellationToken).ConfigureAwait(false);
+                    }
                 }
             }
 
             async Task OnTimeoutAction(Context? context, TimeSpan timeout, Task? task, Exception? exception)
             {
-                if (onTimeoutAction != null)
+                if (hasRetried && onTimeoutAction != null)
```

Comment:
Why do we need to check `hasRetried` here? Wouldn't we still want to call the timeout action if the original request (before retries) times out?

Reply:
Yerp, I've done this to stop logging a misleading message to the user "Could not communicate with Teantcale after 150 seconds. No more retries will be attempted."

As we probably took more than 150 seconds in this scenario as we are still on the initial request and never retried, and saying no more retries will be attempted sounds like we attempted retries that we didn't!!

But as you point out it seems sus to not call on timeout!! I'll look at a better way of logging what really happened to the end user!